### PR TITLE
DOC: add examples to concat

### LINKED ISSUE #4620: Link concat info to concat doc string
Could link http://xarray.pydata.org/en/stable/combining.html#concatenate or add the example to the concat doc string: http://xarray.pydata.org/en/stable/generated/xarray.concat.html

### PR DESCRIPTION
**Pull request by raybellwaves**
- [x] Closes #4620

### CODE REVIEW

File: xarray/core/concat.py [hunk 1]
```diff
@@ -161,6 +161,45 @@ def concat(
     --------
     merge
     auto_combine
+
+    Examples
+    --------
+    >>> da = xr.DataArray(np.random.randn(2, 3),
+    ...                    [("x", ["a", "b"]),
+    ...                    ("y", [10, 20, 30])])
+    >>> xr.concat([da[:, :1], da[:, 1:]], dim="y")
```
Comment (andersy005) [inline xarray/core/concat.py:170]: Wouldn't `isel` be a better alternative here ?


```suggestion
    >>> xr.concat([da.isel(y=slice(0, 1)), da.isel(y=slice(1, None))], dim="y")
```

File: xarray/core/concat.py [hunk 2]
```diff
@@ -161,6 +161,45 @@ def concat(
...
+    <xarray.DataArray (x: 2, y: 3)>
+    array([[-0.99838282, -0.71384379, -0.27952494],
+           [-0.38083294,  0.23861607, -0.32333683]])
+    Coordinates:
+      * x        (x) <U1 'a' 'b'
+      * y        (y) int64 10 20 30
+
+    >>> xr.concat([da[0], da[1]], "x")
```
Comment (andersy005) [inline xarray/core/concat.py:178]:
```suggestion
    >>> xr.concat([da.isel(x=0), da.isel(x=1)], "x")
```

File: xarray/core/concat.py [hunk 3]
```diff
@@ -161,6 +161,45 @@ def concat(
...
+    <xarray.DataArray (x: 2, y: 3)>
+    array([[-1.23581607, -1.00288945,  0.56989523],
+           [-1.03495318,  0.38050723, -0.58316651]])
+    Coordinates:
+      * x        (x) object 'a' 'b'
+      * y        (y) int32 10 20 30
+
+    >>> xr.concat([da[0], da[1]], "new_dim")
```
Comment (andersy005) [inline xarray/core/concat.py:186]:
```suggestion
    >>> xr.concat([da.isel(x=0), da.isel(x=1)], "new_dim")
```

File: xarray/core/concat.py [hunk 4]
```diff
@@ -161,6 +161,45 @@ def concat(
...
+    <xarray.DataArray (new_dim: 2, y: 3)>
+    array([[-1.23581607, -1.00288945,  0.56989523],
+           [-1.03495318,  0.38050723, -0.58316651]])
+    Coordinates:
+        x        (new_dim) <U1 'a' 'b'
+      * y        (y) int32 10 20 30
+    Dimensions without coordinates: new_dim
+
+    >>> xr.concat([da[0], da[1]], pd.Index([-90, -100], name="new_dim"))
```
Comment (andersy005) [inline xarray/core/concat.py:195]:
```suggestion
    >>> xr.concat([da.isel(x=0), da.isel(x=1)], pd.Index([-90, -100], name="new_dim"))
```

File: xarray/core/concat.py [hunk 5]
```diff
@@ -161,6 +161,45 @@ def concat(
     --------
     merge
     auto_combine
+
+    Examples
+    --------
+    >>> da = xr.DataArray(np.random.randn(2, 3),
```
Comment (max-sixty) [inline xarray/core/concat.py:167]: Unfortunately using random will cause the doctests to fail (unless we're careful about the seed, which probably isn't worthwhile vs using arange or similar)

**Review by max-sixty — COMMENTED** [review]
Thanks @raybellwaves ! These make the library much easier for newcombers

File: xarray/core/concat.py (hunk 5 above)
Comment (keewis) [reply xarray/core/concat.py:167]: we do set a seed: https://github.com/pydata/xarray/blob/7152b41fa80a56db0ce88b241fbe4092473cfcf0/conftest.py#L37-L38
but I suspect the expected values were generated without it. It's still better to not rely on randomness at all, especially if the arrays are small enough to be created by hand:
```suggestion
    >>> da = xr.DataArray(
    ...     [[0, 1, 2], [3, 4, 5]],
```

**Comment by raybellwaves** [comment]
Thanks for the feedback.

It was a copy-paste from http://xarray.pydata.org/en/stable/combining.html#concatenate

I'll implement your suggestions in https://github.com/pydata/xarray/blob/master/doc/combining.rst as well.

i.e. use arange for the creation and isel for the index.

**Comment by dcherian** [comment]
Thanks @raybellwaves. Can you address this error?

```
>>>-------------------------------------------------------------------------
Exception in /home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/4645/doc/combining.rst at block ending on line 121
Specify :okexcept: as an option in the ipython:: block to suppress this message
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-16-6d6353d21b81> in <module>
----> 1 xr.Dataset({"a": arr[:-1], "b": arr[1:]})

NameError: name 'arr' is not defined
<<<-------------------------------------------------------------------------
```

**Comment by dcherian** [comment]
Oops one more


```
>>>-------------------------------------------------------------------------
Exception in /home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/4645/doc/combining.rst at block ending on line 191
Specify :okexcept: as an option in the ipython:: block to suppress this message
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-25-4cb6d79e0bd5> in <module>
----> 1 arr.equals(arr.copy())

NameError: name 'arr' is not defined
<<<-------------------------------------------------------------------------
```

File: doc/combining.rst [hunk 6]
```diff
@@ -29,41 +29,42 @@ that dimension:
...
 
-    arr = xr.DataArray(np.random.randn(2, 3), [("x", ["a", "b"]), ("y", [10, 20, 30])])
-    arr[:, :1]
-    # this resembles how you would use np.concatenate
-    xr.concat([arr[:, :1], arr[:, 1:]], dim="y")
+    da = xr.DataArray(np.arange(6).reshape(2, 3), [("x", ["a", "b"]), ("y", [10, 20, 30])])
+    da.isel(y=slice(0, 1)) # same as da[:, :1]
+    # This resembles how you would use np.concatenate:
+    xr.concat([da[:, :1], da[:, 1:]], dim="y")
+    # For more freindly pandas-like indexing you can use:
```
Comment (max-sixty) [inline doc/combining.rst:36]:
```suggestion
    # For more friendly pandas-like indexing you can use:
```

**Review by max-sixty — APPROVED** [review]
Looks really good now. Thanks @raybellwaves

File: xarray/core/concat.py [hunk 7]
```diff
@@ -161,6 +161,48 @@ def concat(
     --------
     merge
     auto_combine
+
+    Examples
+    --------
+    >>> da = xr.DataArray(np.arange(6).reshape(2, 3),
+    ...                   [("x", ["a", "b"]),
+    ...                   ("y", [10, 20, 30])])
```
Comment (keewis) [inline xarray/core/concat.py:169]: I think it would be good to show what the original object looks like. Could you add a line containing `>>> da` here?

**Review by keewis — COMMENTED** [review]
looks good to me. You will have to update the expected output, though. I would also run `pre-commit run --files doc/combining.rst xarray/core/concat.py` once, but that's optional.

**Comment by raybellwaves** [comment]
> I would also run `pre-commit run --files doc/combining.rst xarray/core/concat.py` once, but that's optional.

I'm doing this PR in my windows machine unfortunately, and pre-commit is a pain on windows. Hopefully it's fine as is.

**Comment by keewis** [comment]
> Hopefully it's fine as is.

no worries, as I said that's optional. I can fix that in a different PR (or this one, if you don't mind me pushing to your branch).

**Comment by raybellwaves** [comment]
More than welcome to push to mine.

**Comment by keewis** [comment]
done. Once you update of the doctest output and maybe add a entry to `whats-new.rst` (`Documentation`) this should be ready for merging.

**Comment by raybellwaves** [comment]
I can add to the Whats-new. Not sure how to "update of the doctest output"

**Comment by keewis** [comment]
sorry about that. I meant that the code examples in the doctest still show the results from before when the input data was randomly generated, and this is the reason why the `doctests` CI fails.

For example, this: https://github.com/pydata/xarray/blob/4c8642560679c173c2968a2c4186a474bfe97f2d/xarray/core/concat.py#L178-L185
should be:
```python
     >>> xr.concat([da.isel(y=slice(0, 1)), da.isel(y=slice(1, None))], dim="y") 
     <xarray.DataArray (x: 2, y: 3)> 
     array([[0, 1, 2],
            [3, 4, 5]])
     Coordinates: 
       * x        (x) <U1 'a' 'b' 
       * y        (y) int64 10 20 30 
  
```

**Comment by raybellwaves** [comment]
Not sure if I can do anything for the doctest failing
https://dev.azure.com/xarray/xarray/_build/results?buildId=4491&view=logs&j=e1da92ae-4a54-5c20-3ee1-ef7b06ffcd80&t=5640f7da-bb07-517b-ec85-0b3cfe0ccaa3&l=62
```
    -  * y        (y) int32 10 20 30
    +  * y        (y) int64 10 20 30
```

**Comment by dcherian** [comment]
I think you replace all `int32` by `int64`?

File: doc/whats-new.rst [hunk 8]
```diff
@@ -39,6 +39,9 @@ Documentation
   By `Justus Magin <https://github.com/keewis>`_.
 - start a list of external I/O integrating with ``xarray`` (:issue:`683`, :pull:`4566`).
   By `Justus Magin <https://github.com/keewis>`_.
+- add concat examples and improve combining documentation (:issue:`4620`, :pull:`4645`).
+  By `Ray Bell <https://github.com/raybellwaves>`_ and
+  `Justus Magin <https://github.com/keewis>`_.
```
Comment (keewis) [inline doc/whats-new.rst:44]: you don't really have to add me as co-author, I only reviewed and applied a tool (you can keep it as is if you want, though)

**Review by keewis — APPROVED** [review]
Thanks, @raybellwaves, looks good to me.

File: doc/whats-new.rst (hunk 8 above)
Comment (raybellwaves) [reply doc/whats-new.rst:44]: it's fine

**Comment by mathause** [comment]
Thanks @raybellwaves!

**Pull request merged**